File: finalists/Marcos/challenge_10/10.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import sys

for linea in sys.stdin:
    tokens = linea.split()
    # Inicializamos la pila --> la pila crecerá hacia el final, el último elemento insertado será el último de la lista
    pila = []
    for token in tokens:
        if token.isdigit():
            pila.append(long(token))
        elif token == ".":
            # Imprimir la cabeza de la pila
            print str(pila[-1])
        elif token == "mirror":
            # Eliminar el último elemento y cambiarle el signo
            elem = pila.pop()
            pila.append(-elem)
        elif token == "breadandfish":
            # Replicar el último elemento (cabeza)
            pila.append(pila[-1])
        elif token == "fire":
            # Quitar un elemento de la pila
            pila.pop()
        elif token == "dance":
            # Darle la vuelta a la pila
            pila.reverse()
        elif token == "conquer":
            # Quitar los dos últimos elementos de la pila y hacer el resto (segundo % primero)
            num1 = pila.pop()
            num2 = pila.pop()
            pila.append(num2 % num1)
        elif token == "$":
            # Quitar los dos últimos elementos de la pila y restarlos
            num1 = pila.pop()
            num2 = pila.pop()
            pila.append (num2 - num1)
        elif token == "&":
            # Quitar los dos úlimos elementos de la pila y dividirlos
            num1 = pila.pop()
            num2 = pila.pop()
            pila.append (num2 / num1)
        elif token == "@":
            # Quitar los dos últimos elementos de la pila y sumarlos
            num1 = pila.pop()
            num2 = pila.pop()
            pila.append (num2 + num1)
        elif token == "#":
            # Quitar los dos últimos elementos de la pila y multiplicarlos
            num1 = pila.pop()
            num2 = pila.pop()
            pila.append (num2 * num1)
            
            
        

File: finalists/César/challenge_10/10.java
import sys, hashlib

def process_program(program):
    tokens = program.split(" ")
    stack = []
    for token in tokens:
        if token.isdigit():
            stack.append(int(token))
        elif token == "mirror":
            stack.append(-stack.pop())
        elif token == "breadandfish":
            stack.append(stack[-1])
        elif token == "fire":
            stack.pop()
        elif token == "dance":
            op2 = stack.pop()
            op1 = stack.pop()
            stack.append(op2)
            stack.append(op1)
        elif token == "@":
            stack.append(stack.pop() + stack.pop())
        elif token == "$":
            stack.append(-stack.pop() + stack.pop())
        elif token == "#":
            stack.append(stack.pop() * stack.pop())
        elif token == "&":
            op2 = stack.pop()
            op1 = stack.pop()
            stack.append(op1 / op2)
        elif token == "conquer":
            op2 = stack.pop()
            op1 = stack.pop()
            stack.append(op1 % op2)
        elif token != ".":
            return "UNKNOWN: " + token
    return stack[0]

if __name__ == "__main__":
    for line in sys.stdin:
        print process_program(line.strip())
